src/components/widgets/EditableTextWithSuggestions.vue

```diff
@@ -0,0 +1,206 @@
+<script setup lang="ts">
+// EditableText
+// Hovering over the text shows an edit button
+// Clicking the edit button makes an input field
+// Blurring focus or clicking the save button commits the change
+// Pressing escape or clicking the abort button aborts the change
+
+import { nextTick, ref, unref, watch } from 'vue';
+import DropDown from './DropDown.vue';
+import { isDefined, watchDebounced } from '@vueuse/core';
+
+const props = withDefaults(defineProps<{
+  modelValue: string | null,
+  readOnly?: boolean,
+  editing?: boolean,
+  placeholder?: string,
+  error?: string | null,
+  debounce?: number,
+}>(), {
+  debounce: 500,
+});
+
+const emit = defineEmits<{
+  'update:modelValue': [value: string | null],
+  'update:query': [query: string],
+}>();
+
+const elInput = ref<HTMLInputElement>();
+const editing = ref<boolean>(props.editing ?? false);
+
+const editedModelValue = ref<string | null>(unref(props.modelValue));
+
+watch(() => props.modelValue, () => {
+  editedModelValue.value = props.modelValue;
+});
+
+watchDebounced(editedModelValue, async () => {
+  if (!isDefined(editedModelValue)
+    || editedModelValue.value === ''
+    || editedModelValue.value === props.modelValue) {
+    return;
+  }
+
+  emit('update:query', editedModelValue.value);
+}, { debounce: props.debounce });
+
+function beginEdit() {
+  editing.value = true;
+  nextTick(() => {
+    elInput.value?.focus();
+  });
+}
+
+function commit() {
+  editing.value = false;
+  emit('update:modelValue', editedModelValue.value);
+}
+
+function abort() {
+  editing.value = false;
+  editedModelValue.value = props.modelValue;
+}
+
+function onBlur(e: FocusEvent | null) {
+  if (e == null || e.currentTarget == null) {
+    commit();
+    return;
+  }
+
+  if (
+    e.relatedTarget !== e.currentTarget
+    && !(e.currentTarget as Node)?.contains(e.relatedTarget as Node)
+  ) {
+    commit();
+  }
+
+}
+</script>
+
+<template>
+  <!-- for invalid-tooltip positioning -->
+  <div class="position-relative col-9 osim-editable-field osim-text">
+    <!--<Transition name="flash-bg" :duration="2000">-->
+    <Transition name="flash-bg">
+      <div
+        v-show="!editing"
+        class="osim-editable-text"
+        :tabindex="readOnly ? -1 : 0"
+        @focus="beginEdit"
+      >
+        <span
+          class="osim-editable-text-value"
+          :class="{
+            'form-control': !readOnly,
+            'is-invalid': error != null
+          }"
+        >{{ modelValue === '' ? placeholder : modelValue }}</span>
+        <!--if a button is inside a label, clicking the label clicks the button?-->
+        <button
+          v-if="!readOnly"
+          type="button"
+          class="osim-editable-text-pen input-group-text"
+          tabindex="-1"
+          @click="beginEdit"
+        ><i class="bi bi-pencil"></i></button>
+        <slot name="buttons-out-of-editing-mode" v-bind="{ commit, onBlur }"></slot>
```

Comment:
I don't understand why we do require two different slots `buttons-out-of-editing-mode` and `buttons-in-editing-mode`

Reply:
It was like that, is part of the `EditableText` widget, I copied the code to add the suggestions part so the API is the same on both